File: darknet_ros/scripts/tokamak.py
#!/usr/bin/env python3
"""
Node to synchronise incoming Kinect RGB camera streams to relay to Darknet

Created 08/13/2023 by Sasha Lew
"""
from message_filters import ApproximateTimeSynchronizer, Subscriber
import rospy

from sensor_msgs.msg import Image

class Tokamak:
    """Feedforward synchronised master and sub Kinect RGB images."""
    def __init__(self):
        self.image_fusion_pub = rospy.Publisher("/kinect/combined/image_raw", Image, queue_size=15)
        master_image = Subscriber("/kinect/master/rgb/image_raw", Image)
        sub_image = Subscriber("/kinect/sub/rgb/image_raw", Image)
        ts = ApproximateTimeSynchronizer([master_image, sub_image], queue_size=15, slop=0.1)
        ts.registerCallback(self.image_callback)

    def image_callback(self, master_im, sub_im):
        self.image_fusion_pub.publish(master_im)
        # NOTE: sleeping between publishing allows Darknet threading to fetch both images
        rospy.sleep(0.03)
        self.image_fusion_pub.publish(sub_im)

def main():
    try:
        rospy.init_node("tokamak")
        Tokamak()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass

if __name__ == "__main__":
    main()

File: darknet_ros/src/tokamak.cpp
#include <ros/time.h>
#include "darknet_ros/tokamak.h"

Tokamak::Tokamak(const ros::NodeHandle& nh) : nh_(nh),
  it_(nh_),
  master_im_sub_(it_, "/kinect/master/rgb/image_raw", 3),
  sub_im_sub_(it_, "/kinect/sub/rgb/image_raw", 3),
  sync_(SerialImagePolicy(15), master_im_sub_, sub_im_sub_)
{
  image_fusion_pub_ = it_.advertise("/kinect/combined/image_raw", 6);
  sync_.registerCallback(boost::bind(&Tokamak::filterCallback, this, _1, _2));
}

void Tokamak::filterCallback(const sensor_msgs::ImageConstPtr& master_im, const sensor_msgs::ImageConstPtr& sub_im)
{
  image_fusion_pub_.publish(master_im);
  ros::Duration(0.03).sleep();
  image_fusion_pub_.publish(sub_im);
}

int main(int argc, char** argv)
{
  ros::init(argc, argv, "tokamak");
  ros::NodeHandle nh("~");

  Tokamak tokamak(nh);

  ros::AsyncSpinner spinner(4);
  spinner.start();
  ros::waitForShutdown();
}
